Flag for manual review design doc

bkjohnson opened the pull request:
This is a draft design doc for what we've been calling the automated "flag for manual review" process.  It is not _strictly_ a design doc since I ended up creating the solution ahead of time as I was experimenting with Github's API and Circle's CI config, but I tried to capture some design decisions and open questions.

All feedback welcome.

[Here is an earlier issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2521) which describes some of the goals. I will try to work that into the design doc somehow.

bkjohnson committed:
Create flag-for-manual-review.md

More revisions will follow

bkjohnson committed:
Add some more overview info

bkjohnson committed:
Add some specifics

bkjohnson committed:
Add some more details

bkjohnson committed:
Remove some remplate info

bkjohnson committed:
Merge branch 'master' into bkjohnson/flag-for-manual-review-doc

bkjohnson committed:
Add more info on what triggers a manual review

bkjohnson committed:
Add a caveat

bkjohnson committed:
Move into design-docs folder

rianfowler commented on platform/engineering/design-docs/flag-for-manual-review.md:
```diff
@@ -0,0 +1,113 @@
+The goal is to improve our review process which by adding automated processes which will alert a review team when a manual review is required.
+These changes are meant to work alongside the Codeowners changes which will give each app team primary responsibility for reviewing their code.
+
+Non-goals: Creating automation which will absolve VFS teams of their review responsibility.
```
```suggestion
```

I don't think this strikes the right tone. We want to automate the easy stuff so VFS teams can focus their reviews on the harder problems. For the things the platform must enforce but can't automate, we're using this trigger.

bkjohnson replied:
That's fair - I'll remove it.

rianfowler commented on platform/engineering/design-docs/flag-for-manual-review.md:
```diff
@@ -0,0 +1,113 @@
+
+### Logging
+The script will log to the console:
+- Whether or not the branch ir is running on has an associated PR
```
```suggestion
- Whether or not the branch it is running on has an associated PR
```

rianfowler commented on platform/engineering/manual-review-triggers.md:
```diff
@@ -0,0 +1,30 @@
+There are situations where it makes sense to disable certain rules, but those will have 
+to be evaluated on a case-by-case basis.
+
+### Adding icons
```
This is cool

rianfowler reviewed (approved):
cc @gunsch @johnpaulashenfelter

bkjohnson commented on platform/engineering/code_review_guidelines.md:
```diff
@@ -15,9 +15,9 @@ Information on how VSP uses code owners can be found [here](code-owners.md).
-3. Each additional round of changes will be reviewed within one business day.
-4. When all review comments have been resolved, the PR can be merged into the master branch for deployment.
+2. If your PR triggers any [additional automated checks](./manual-review-triggers), a bot will leave a comment and request a manual review from the **frontend-review-group**
+    - Wait for the review group to provide a manual review - it should be completed within one business day.
```
Unfortunately [the API doesn't let us do that](https://developer.github.com/v3/pulls/review_requests/#create-a-review-request).  All we can do is say who we want to request a review from.

However, I included a section talking about other ways that we could enforce that they wait for a manual review.

bkjohnson committed:
Fix typo

Co-Authored-By: Rian Fowler <rian.fowler@adhocteam.us>

bkjohnson committed:
Remove harsh-sounding non-goal

Co-Authored-By: Rian Fowler <rian.fowler@adhocteam.us>

bkjohnson committed:
Be more succinct about review turnaround

Co-Authored-By: Rian Fowler <rian.fowler@adhocteam.us>

gunsch commented on platform/engineering/manual-review-triggers.md:
```diff
@@ -0,0 +1,30 @@
+### Sentry calls
+
+We want to ensure that anything logged to Sentry will be useful (don't log an entire request response if all we care
+about is an error code) and will not contain PII.
```
```suggestion
We review logs to Sentry to ensure that any new logs will be useful and will not contain PII.

Examples:
- Don't log an entire request response if all that's important is an error code
- Don't log user-identifying information such as names, emails, or user IDs
```

gunsch commented on platform/engineering/manual-review-triggers.md:
```diff
@@ -0,0 +1,30 @@
+
+We have various ESLint rules in place to help improve code quality.
+There are situations where it makes sense to disable certain rules, but those will have 
+to be evaluated on a case-by-case basis.
```
```suggestion
Disabling eslint rules will be evaluated on a case-by-case basis.
```

gunsch commented on platform/engineering/manual-review-triggers.md:
```diff
@@ -0,0 +1,30 @@
+
+We use fontawesome as a dependency which encourages icons to be added with the `<i>` tag.
+Sometimes an icon is used purely as decoration, but other times it is used to convey meaning to the user.
+We want to ensure that if an icon is being used semantically, those semantics are also conveyed to a screen reader.
```
```suggestion
We use Font Awesome as a dependency, which uses the `<i>` tag for adding icons. Sometimes an icon is used purely as decoration, but other times it is used to convey meaning to the user.

We review to ensure that whenever an icon is being used semantically, those semantics are also conveyed to a screen reader.
```

bkjohnson committed:
Apply suggestions from code review

Co-Authored-By: Andrew Gunsch <andrew.gunsch@adhocteam.us>

bkjohnson committed:
Update security concerns section

bkjohnson committed:
Add detail on how to handle the token securely

bkjohnson committed:
Add more info on environment variables

Pulled from PR description in https://github.com/department-of-veterans-affairs/vets-website/pull/11670

bkjohnson committed:
Use file extension in link

Co-Authored-By: Andrew Gunsch <andrew.gunsch@adhocteam.us>

bkjohnson committed:
Move design doc out of design-docs folder

That seems to be just for the template

bkjohnson committed:
Fix typo in environment variables link

bkjohnson commented on platform/engineering/design-docs/flag-for-manual-review.md:
```diff
@@ -0,0 +1,112 @@
+
+### Debugging
+
+If a user sets up the proper environment variables locally, the script can be run in a local environment and the logging output can be viewed.  Otherwise we will rely on viewing CircleCI logs.
```
I put the environment variables in the same doc with the list of triggers and linked to it.  I'm not convinced that that is the best place,  so that may change as I clean things up.

bkjohnson committed:
Add detail on discovering when process is ignored

bkjohnson committed:
Add repo and file description to code location

bkjohnson commented on platform/engineering/design-docs/flag-for-manual-review.md:
```diff
@@ -0,0 +1,112 @@
+1. Getting approval from a member of the VSP *frontend-review-group*
+
+The final approval from the review group is often not helpful, and it slows the process down.
+This frustration often leads to VFS teams ignoring this process and they will directly ask for a VSP review before their PR is ready.
```
I avoided that because that felt like part of the solution, and the template said to [keep that out of the background](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/design-doc-template.md#background).  I'll try to phrase it in a way that keeps the emphasis on the problem and not the solution.

bkjohnson committed:
Be more specific about testing plan

bkjohnson committed:
Add a bit of background on codeowners changes

bkjohnson committed:
Add detail on how the Github API is used

bkjohnson committed:
Fix ref case

gunsch commented on platform/engineering/flag-for-manual-review.md:
```diff
@@ -39,24 +41,40 @@ This script will look for anything that should trigger a manual review by the VS
 ### Detailed Design
 
-We will use a script that can be run from a job in CircleCI.
+We will use a nodeJS script that can be run from a job in CircleCI.
```
```suggestion
We will use a Node.js script that can be run from a job in CircleCI.
```

gunsch commented on platform/engineering/flag-for-manual-review.md:
```diff
@@ -76,7 +94,11 @@ No other context is included.
 
-We should make sure that the Github bot auth token is kept secure and it should not be associated with an individual's account.  The sole communicators in the system are CircleCI and Github.  This isn't really a security concern, but more of a reliability one: if one of those services goes down we do not have a fallback plan.
+We should make sure that the Github bot auth token is kept secure and it should not be associated with an individual's account.
+The current strategy for storing this token is as an [environment variable in CircleCI](https://ui.circleci.com/settings/project/github/department-of-veterans-affairs/vets-website/environment-variables) called `GITHUB_TOKEN`.
```
From the [CircleCI docs](https://circleci.com/docs/2.0/env-vars/):

> Note: secret masking will only prevent the value of the environment variable from appearing in your build output. The value of the environment variable is still accessible to users debugging builds with SSH.

who has SSH access to these builds?

bkjohnson replied:
My understanding is that anyone who has an SSH public key on Github will be able to connect. I re-ran a job with SSH and this is what I see:

![image](https://user-images.githubusercontent.com/2008881/76005767-5621aa00-5ec0-11ea-9641-e65dfb465e9f.png)

So anytime we deliberately kick off a job with SSH access, we should be able to see whoever connects to it.

gunsch replied:
_anyone_ with an SSH public key? So, that sounds to me like the GitHub bot auth token can be accessed by anyone?

bkjohnson replied:
I found a more specific answer in [the docs](https://circleci.com/docs/2.0/ssh-access-jobs/):

> In order to SSH into a CircleCI build, the username must be one which has access to the project being built!

When I try to go to [the project website](https://app.circleci.com/pipelines/github/department-of-veterans-affairs/vets-website) in a private window it blocks me and forces me to authenticate with Github or Bitbucket.

I created a fake github account to test this and was able to login and view the project in Circle, but I can't even look at the project settings. I get errors when I try to follow the project, and I can't see the list of branches either.  The option to trigger an SSH build is disabled as well:

![image](https://user-images.githubusercontent.com/2008881/76033868-27242c00-5ef2-11ea-8d9e-52f26d3193d5.png)

After seeing all of this, I think that project access in circle requires an account to be a part of the Github organization that the repo is associated with. I've looked around at various settings to try to get a more definitive answer, but I'm not an organization admin in Circle so there are certain things I can't see. I wonder if [this page](https://circleci.com/gh/organizations/department-of-veterans-affairs/settings#projects) would be helpful.

gunsch replied:
https://discuss.circleci.com/t/disable-ssh/31057 has a couple key things:

> Any user with write capabilities will also have SSH access. [...] any user with write permissions could theoretically also expose secrets via commands issues through the config.yml file.

Related question to investigate: how often do we hand out `write` permission on vets-website/vets-api, or do we generally keep people's access to `read`?

> A good first step is to ensure your deploy keys only have read access to your repos.

This made me wonder, what kinds of access does our bot token have? Can we keep that as minimal as possible? (read-only + posting comments?).

Other possible idea: it sounds like [contexts](https://circleci.com/docs/2.0/contexts/#restricting-a-context) might be useful here.

bkjohnson replied:
Do you know who gives `write` access to people?

Our bot token has the [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes).  I believe that this is the least permissions that we can give it in order for it to do its job, but it does include write access for our public repos.

> Limits access to public repositories. That includes read/write access to code, commit statuses, repository projects, collaborators, and deployment statuses for public repositories and organizations.

Here is a [relevant Slack thread](https://dsva.slack.com/archives/CJYRZK2HH/p1581623319018100?thread_ts=1581620875.013600&cid=CJYRZK2HH).

Contexts are something that I haven't explored deeply.  I'm interested to see how we could use them, but I don't have permissions to manage security groups so I'm a bit limited.

gunsch replied:
> Do you know who gives `write` access to people?

VA has an entire GitHub admin group. you can also talk to @ricetj about longer-term strategy for ownership on these repos, or (more immediately helpful) see if you can track down an admin for `vets-website` and `vets-api` to see how we've handed out write access so far. it's possible that most people just have `read` access and that this isn't a concern, but `write` access is typically needed for managing issues, which might mean we've handed it out more broadly.

> Our bot token has the [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes). I believe that this is the least permissions that we can give it in order for it to do its job, but it does include write access for our public repos.

hmm, this set of information is making me more and more concerned 😬 

what does the bot need from `public_repo` that it can't get from `(no scope)`?

rianfowler replied:
FWIW- Write access is needed by a GitHub team for codeowners to trigger so everyone has write access to the repo. 

All of the FE group on VSP has admin access. 

Generally everyone has `write` access.

gunsch commented on platform/engineering/manual-review-triggers.md:
```diff
@@ -14,17 +14,43 @@ Any of the following items if found in a PR diff will prompt a manual review
 We have various ESLint rules in place to help improve code quality.
-There are situations where it makes sense to disable certain rules, but those will have 
-to be evaluated on a case-by-case basis.
+Disabling eslint rules will be evaluated on a case-by-case basis.
```
```suggestion
Disabling ESLint rules will be evaluated on a case-by-case basis.
```

gunsch commented on platform/engineering/flag-for-manual-review.md:
```diff
@@ -0,0 +1,144 @@
+**Author(s):** Brooks Johnson (brooks@adhoc.team)
+**Last Updated:** February 20, 2020  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** _Rian Fowler_ \[ \], _Andrew Gunsch_ \[ \], ...  
```
add @drorva to this list, he's asked to be on all design docs for now

```suggestion
**Approvers:** Rian Fowler \[ \], Andrew Gunsch \[X\], Dror Matalon \[ \]
```

gunsch reviewed (approved):
Design doc location: maybe back in the `design-docs` folder, and prepended with a date? we don't really have a convention yet for where these live, but the `engineering` directory is pretty convoluted already with a bunch of different files.

Couple small comments here, but otherwise this looks pretty good! adding my approval (but make sure to get others' also before submitting).

bkjohnson committed:
Apply suggestions from code review

Co-Authored-By: Andrew Gunsch <andrew.gunsch@adhocteam.us>

bkjohnson committed:
Remove manual-review-triggers doc

This was pulled out and added in this PR:
- https://github.com/department-of-veterans-affairs/va.gov-team/pull/6592

bkjohnson committed:
Mark as approved
